jenkins/check_projects.py

```diff
@@ -39,190 +41,393 @@
-null = open('/dev/null', 'w')
-
-# Not thread safe, acquire lock before accessing.
-completed = set()  # Projects with no updated needed
-updated = set()  # Projects which changed
-errors = set()  # Projects which failed to fully update
-helpers = {}  # People with owners rights to update error projects
+logger = logging.getLogger('check_project')
 
 DEFAULT = {
-    'roles/editor': [
+    'IAM': {
+      'roles/editor': [
         'serviceAccount:kubekins@kubernetes-jenkins.iam.gserviceaccount.com',
-        'serviceAccount:pr-kubekins@kubernetes-jenkins-pull.iam.gserviceaccount.com',
-    ],
+        'serviceAccount:pr-kubekins@kubernetes-jenkins-pull.iam.gserviceaccount.com'],
+    },
+    'EnableVMZonalDNS': False,
 }
 
 
+class RateLimitedExec(object):
+  def __init__(self):
+    self._semaphore = threading.Semaphore(10)  # 10 concurrent gcloud calls
+
+  def check_output(self, *args, **kwargs):
+    with self._semaphore:
+      return subprocess.check_output(*args, **kwargs)
+
+  def call(self, *args, **kwargs):
+    with self._semaphore:
+      return subprocess.call(*args, **kwargs)
+
+
+class Results(object):
+  class Info(object):
+    def __init__(self):
+      self.updated = False
+      self.errors = [] # errors is list of strings describing the error context
+      self.helpers = []  # People with owners rights to update error projects
+
+  def __init__(self):
+    self._lock = threading.Lock()
+    self._projects = {} # project -> Info
+
+  @property
+  def errors(self):
+    """
+    |return| set of projects that have errors.
+    """
+    return set([key for key in self._projects if self._projects[key].errors])
+
+  @property
+  def projects(self):
+    return self._projects
+
+  def reportProject(self, project):
+    with self._lock:
+      self._ensureInfo(project)
+
+  def reportError(self, project, err):
+    with self._lock:
+      info = self._ensureInfo(project)
+      info.errors.append(err)
+
+  def reportUpdated(self, project):
+    with self._lock:
+      info = self._ensureInfo(project)
+      info.updated = True
+
+  def addHelper(self, project, helpers):
+    with self._lock:
+      info = self._ensureInfo(project)
+      info.helpers = helpers
+
+  def counts(self):
+    """
+    |return| count of (total, updated, error'ed) projects.
+    """
+    with self._lock:
+      total, updated, errored = (0, 0, 0)
+      for info in self._projects.values():
+        total += 1
+        if info.updated:
+          updated += 1
+        if info.errors:
+          errored += 1
+      return (total, updated, errored)
+
+  def _ensureInfo(self, project):
+    if project not in self._projects:
+      info = Results.Info()
+      self._projects[project] = info
+    return self._projects[project]
+
+
+def runThreadsToCompletion(threads):
+  """
+  Runs the given list of threads to completion.
+  """
+  for thread in threads:
+    thread.start()
+  for thread in threads:
+    thread.join()
+
+
+def parseArgs():
```

Comment:
I think this would be easier to find if it is near the bottom or else at the top